gc/default/default.c

```diff
@@ -1614,6 +1595,45 @@ rb_gc_impl_object_id(void *objspace_ptr, VALUE obj)
     return id;
 }
 
+VALUE
+rb_gc_impl_object_id_to_ref(void *objspace_ptr, VALUE object_id)
+{
+    rb_objspace_t *objspace = objspace_ptr;
+
+    VALUE obj;
+    if (st_lookup(objspace->id_to_obj_tbl, object_id, &obj) &&
+            !rb_gc_impl_garbage_object_p(objspace, obj)) {
+        return obj;
+    }
+
+    if (rb_funcall(object_id, rb_intern(">="), 1, ULL2NUM(objspace->next_object_id))) {
+        rb_raise(rb_eRangeError, "%+"PRIsVALUE" is not id value", rb_funcall(object_id, rb_intern("to_s"), 1, INT2FIX(10)));
+    }
+    else {
+        rb_raise(rb_eRangeError, "%+"PRIsVALUE" is recycled object", rb_funcall(object_id, rb_intern("to_s"), 1, INT2FIX(10)));
+    }
+}
+
+st_data_t
+rb_gc_impl_object_id_move(void *objspace_ptr, VALUE dest, VALUE src)
+{
+   /* If the source object's object_id has been seen, we need to update
+    * the object to object id mapping. */
+   st_data_t id = 0;
+   rb_objspace_t *objspace = objspace_ptr;
+
+   unsigned int lev = rb_gc_vm_lock();
+   st_data_t key = (st_data_t)src;
+   if (!st_delete(objspace->obj_to_id_tbl, &key, &id)) {
+       rb_bug("gc_move: object ID seen, but not in mapping table: %s", rb_obj_info(src));
+   }
+   st_insert(objspace->obj_to_id_tbl, (st_data_t)dest, id);
+   st_insert(objspace->id_to_obj_tbl, id, (st_data_t)dest);
+   FL_UNSET_RAW(src, FL_SEEN_OBJ_ID);
+   rb_gc_vm_unlock(lev);
```

Comment:
This is going to break `ObjectSpace._id2ref`. Is there a reason why we want the moved object to have the same object ID? It's a new object, so IMO it should have a different object ID.

Reply:
> This is going to break `ObjectSpace._id2ref`.

How so?

> Is there a reason why we want the moved object to have the same object ID? It's a new object,

Semantically, it's not a new object. It's the same object but moved to another ractor.

I'm fine with not transferring the `object_id`, the previous code was doing it AFAICT, but it feels like the right thing to do.

Reply:
Ah I misread the code. I didn't see that you removed the original object from the `obj_to_id_tbl` table. This makes sense now 👍